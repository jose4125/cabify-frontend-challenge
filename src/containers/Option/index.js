import React from 'react';

export class Option extends React.Component {
  constructor(props) {
    super(props);
    this.OptionRefSelected = React.createRef();
  }

  componentDidMount() {
    if (this.props.counter === this.props.index) {
      this.props.handleOptionselectedElement(this.OptionRefSelected);
    }
  }
  componentDidUpdate(prevProps) {
    if (
      prevProps.counter !== this.props.index &&
      this.props.counter === this.props.index
    ) {
      this.props.handleOptionselectedElement(this.OptionRefSelected);
    }
  }
  render() {
    return (
      <li
        ref={this.OptionRefSelected}
        className={`dropdown__item ${
          this.props.countryCode.dial_code === this.props.countryCodeSelected
            ? 'selected'
            : ''
        }`}
        onClick={event => {
          this.props.setCountrySelected(this.props.countryCode.dial_code);
          this.props.handleOpenCloseDropdown(false);
        }}
      >
        <div className="dropdown__country">
          <span className={`flag ${this.props.countryCode.code}`} />
          <p className="dropdown__country-name">
            {this.props.countryCode.name}
          </p>
        </div>
        {this.props.countryCode.dial_code}
      </li>
    );
  }
}

export default Option;
