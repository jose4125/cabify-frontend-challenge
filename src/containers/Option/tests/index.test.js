import React from 'react';
import { shallow, mount } from 'enzyme';

import Option from '../index';

describe('Option', () => {
  const mockHandleOpenCloseDropdown = jest.fn();
  const mockSetCountrySelected = jest.fn();
  const mockHandleOptionselectedElement = jest.fn();
  let props;
  let prevProps;
  let wrapper;

  beforeEach(() => {
    props = {
      key: 'United State',
      index: 2,
      counter: 0,
      countryCode: { dial_code: '+34', name: 'United State', code: 'US' },
      countryCodeSelected: '+34',
      handleOpenCloseDropdown: mockHandleOpenCloseDropdown,
      setCountrySelected: mockSetCountrySelected,
      handleOptionselectedElement: mockHandleOptionselectedElement
    };

    prevProps = {
      counter: 0
    };

    wrapper = shallow(<Option {...props} />);
  });

  afterEach(() => {
    wrapper.unmount();
  });

  it('should render with out crashing', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should call handleOptionselectedElement when is mounted', () => {
    props.counter = 2;
    wrapper = mount(<Option {...props} />);

    expect(mockHandleOptionselectedElement).toHaveBeenCalledTimes(1);
  });

  it('should call handleOptionselectedElement when is update', () => {
    props.counter = 1;
    props.index = 1;
    wrapper.instance().componentDidUpdate(prevProps);

    expect(mockHandleOptionselectedElement).toHaveBeenCalledTimes(1);
  });

  it('should have a selected class', () => {
    expect(wrapper.prop('className')).toContain('selected');
  });

  it('should have a span with country code flag class', () => {
    expect(wrapper.find('.flag').prop('className')).toContain(
      props.countryCode.code
    );
  });

  it('should have a country name text', () => {
    expect(wrapper.find('.dropdown__country-name').text()).toBe(
      props.countryCode.name
    );
  });

  it('should have a country dial code text', () => {
    expect(wrapper.text()).toBe(
      `${props.countryCode.name}${props.countryCode.dial_code}`
    );
  });

  it('should call setCountrySelected and handleOpenCloseDropdown when is clicked', () => {
    wrapper.simulate('click');
    expect(mockSetCountrySelected).toHaveBeenCalledWith(
      props.countryCode.dial_code
    );

    expect(mockSetCountrySelected).toHaveBeenCalledTimes(1);

    expect(mockHandleOpenCloseDropdown).toHaveBeenCalledWith(false);
    expect(mockHandleOpenCloseDropdown).toHaveBeenCalledTimes(1);
  });
});

describe('Option no selected', () => {
  let props;
  let wrapper;

  beforeEach(() => {
    props = {
      key: 'United State',
      index: 2,
      counter: 0,
      countryCode: { dial_code: '+34', name: 'United State', code: 'US' },
      countryCodeSelected: '+93'
    };

    wrapper = shallow(<Option {...props} />);
  });

  it('should not have a selected class', () => {
    expect(wrapper.prop('className')).not.toContain('selected');
  });
});
