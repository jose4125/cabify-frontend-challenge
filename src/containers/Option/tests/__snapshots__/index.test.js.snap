// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Option should render with out crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Option
    counter={0}
    countryCode={
      Object {
        "code": "US",
        "dial_code": "+34",
        "name": "United State",
      }
    }
    countryCodeSelected="+34"
    handleOpenCloseDropdown={[MockFunction]}
    handleOptionselectedElement={[MockFunction]}
    index={2}
    setCountrySelected={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="dropdown__country"
        >
          <span
            className="flag US"
          />
          <p
            className="dropdown__country-name"
          >
            United State
          </p>
        </div>,
        "+34",
      ],
      "className": "dropdown__item selected",
      "onClick": [Function],
    },
    "ref": Object {
      "current": null,
    },
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <span
              className="flag US"
            />,
            <p
              className="dropdown__country-name"
            >
              United State
            </p>,
          ],
          "className": "dropdown__country",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "flag US",
            },
            "ref": null,
            "rendered": null,
            "type": "span",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "United State",
              "className": "dropdown__country-name",
            },
            "ref": null,
            "rendered": "United State",
            "type": "p",
          },
        ],
        "type": "div",
      },
      "+34",
    ],
    "type": "li",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="dropdown__country"
          >
            <span
              className="flag US"
            />
            <p
              className="dropdown__country-name"
            >
              United State
            </p>
          </div>,
          "+34",
        ],
        "className": "dropdown__item selected",
        "onClick": [Function],
      },
      "ref": Object {
        "current": null,
      },
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <span
                className="flag US"
              />,
              <p
                className="dropdown__country-name"
              >
                United State
              </p>,
            ],
            "className": "dropdown__country",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "flag US",
              },
              "ref": null,
              "rendered": null,
              "type": "span",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "United State",
                "className": "dropdown__country-name",
              },
              "ref": null,
              "rendered": "United State",
              "type": "p",
            },
          ],
          "type": "div",
        },
        "+34",
      ],
      "type": "li",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": false,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
